sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/storage/block/FileCompactBlockRepository.kt

```diff
@@ -0,0 +1,148 @@
+            check(tmpFile.finalizeFile()) {
+                "Failed to finalize file: ${tmpFile.absolutePath}"
+            }
+            count++
+
+            if (index % ZcashSdk.BLOCKS_METADATA_BUFFER_SIZE == 0) {
+                // write blocks metadata
+                rustBackend.writeBlockMetadata(metaDataBuffer.toTypedArray())
+                metaDataBuffer.clear()
+            }
+        }
+
+        if (metaDataBuffer.isNotEmpty()) {
+            rustBackend.writeBlockMetadata(metaDataBuffer.toTypedArray())
+        }
+
+        return count
+    }
+
+    override suspend fun rewindTo(height: BlockHeight) {
+        rustBackend.rewindToHeight(height)
+    }
+
+    override suspend fun close() {
+        // nothing to do
+    }
+
+    companion object {
+        private var dbInitialized = false
+
+        suspend fun new(
+            zcashNetwork: ZcashNetwork,
+            rustBackend: RustBackend
+        ): FileCompactBlockRepository {
+            twig("${rustBackend.fsBlockDbRoot.absolutePath} \n  ${rustBackend.dataDbFile.absolutePath}")
+            // create cache directories
+            val blocksDirectory =
+                File(rustBackend.fsBlockDbRoot, ZcashSdk.BLOCKS_DOWNLOAD_DIRECTORY).apply { mkdirsSuspend() }
+
+            dbInitialized =  rustBackend.initBlockMetaDb() == 0
+
+            return FileCompactBlockRepository(zcashNetwork, blocksDirectory, rustBackend)
+        }
+    }
+}
+
+private val CompactBlock.outputs: Pair<Int, Int>
+    get() = vtxList.map { compactTx ->
+        compactTx.outputsCount to compactTx.actionsCount
+    }.reduce { partialResult, txOutputActionPair ->
+        (partialResult.first + txOutputActionPair.first) to (partialResult.second + txOutputActionPair.second)
+    }
+
+private fun CompactBlock.toJniMetaData() =
+    JniBlockMeta(
+        height = height,
+        hash = hash.toByteArray(),
+        time = time.toLong(),
+        saplingOutputsCount = outputs.first.toLong(),
+        orchardOutputsCount = outputs.second.toLong()
+    )
+
+@VisibleForTesting
+private fun CompactBlock.createFilename() = "$height-${hash.toByteArray().toHex()}${
+    ZcashSdk
+        .BLOCK_FILENAME_SUFFIX
+}"
+
+@VisibleForTesting
+internal suspend fun CompactBlock.createTemporaryFile(blocksDirectory: File): File {
+    val tempFileName = "${createFilename()}${ZcashSdk.TEMPORARY_FILENAME_SUFFIX}"
+    val tmpFile = File(blocksDirectory, tempFileName)
+
+    if (tmpFile.existsSuspend()) {
+        tmpFile.deleteRecursivelySuspend()
+    }
+    tmpFile.createNewFileSuspend()
+
+    return tmpFile
+}
+
+@VisibleForTesting
+internal fun File.getBlockHeight() = name.split("-")[0].toLong()
+
+@VisibleForTesting
+internal suspend fun File.finalizeFile(): Boolean {
+    // rename the file
+    val newFile = File(absolutePath.dropLast(ZcashSdk.TEMPORARY_FILENAME_SUFFIX.length))
+    return renameToSuspend(newFile)
+}
+
+private fun ByteArray.toHex(): String {
```

Comment:
The SDK should already have a function that already turns a ByteArray to the hex representation of a transactionId so this shouldn't be needed